File: 백준/Gold/2205. 저울 추 만들기/저울 추 만들기.py
# 큰 수 일수록 2의 거듭제곱이 가능한 경우가 별로 없어서 큰 수부터 처리
N = int(input())

check = []
idx = 2
while True:
    if idx <= (2 * N):
        check.append(idx)
    else:
        break
    idx *= 2

weight = [0] * (N+1)
weight[0] = 1

for i in range(N, 0, -1):
    for k in check:
        if k > i and k - i <= N:
            if weight[i] == 0 and weight[k-i] == 0:
                weight[i] = k-i
                weight[k-i] = i

for i in range(1, N+1):
    print(weight[i])

File: 백준/Gold/2205. 저울 추 만들기/저울 추 만들기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        ArrayList<Integer> check = new ArrayList<>();

        int idx = 2;
        while (true) {
            if (idx <= 2 * N) {
                check.add(idx);
            } else {
                break;
            }
            idx *= 2;
        }

        int[] weight = new int[N+1];

        for (int i = N; i > 0; i--) {
            for (int k : check) {
                if (k - i <= N && k > i) {
                    if (weight[i] == 0 && weight[k-i] == 0) {
                        weight[i] = k-i;
                        weight[k-i] = i;
                    }
                }
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= N; i++) {
            sb.append(weight[i]).append("\n");
        }
        System.out.print(sb);
    }
}
